File: app/src/main/java/com/bluzi/babymonitor/monitor/AlarmSchedule.kt
package com.bluzi.babymonitor.monitor

import com.bluzi.babymonitor.data.Settings

// ALRM-7: when the noise alarm is armed. Pure — time injected as minutes-of-day.

data class AlarmSchedule(
    val windowed: Boolean,
    val startMinutes: Int = 0,
    val endMinutes: Int = 0,
) {
    /** Is the alarm armed at [minutesOfDay] (0..1439)? Windows may cross midnight. */
    fun isActive(minutesOfDay: Int): Boolean = when {
        !windowed -> true
        startMinutes == endMinutes -> true // degenerate window = always
        startMinutes < endMinutes -> minutesOfDay in startMinutes until endMinutes
        else -> minutesOfDay >= startMinutes || minutesOfDay < endMinutes // crosses midnight
    }

    companion object {
        fun from(s: Settings): AlarmSchedule = AlarmSchedule(
            windowed = s.alarmScheduleMode == Settings.SCHEDULE_WINDOW,
            startMinutes = s.alarmWindowStartMinutes,
            endMinutes = s.alarmWindowEndMinutes,
        )
    }
}

File: windows/src/BabyMonitor.Core/Monitor/AlarmSchedule.cs
using BabyMonitor.Core.Data;

namespace BabyMonitor.Core.Monitor;

/// <summary>ALRM-7: when the noise alarm is armed. Pure — time injected as minutes-of-day.</summary>
public sealed record AlarmSchedule(bool Windowed, int StartMinutes = 0, int EndMinutes = 0)
{
    /// <summary>Is the alarm armed at <paramref name="minutesOfDay"/> (0..1439)? Windows may cross midnight.</summary>
    public bool IsActive(int minutesOfDay)
    {
        if (!Windowed)
        {
            return true;
        }

        if (StartMinutes == EndMinutes)
        {
            return true; // degenerate window = always
        }

        return StartMinutes < EndMinutes
            ? minutesOfDay >= StartMinutes && minutesOfDay < EndMinutes
            : minutesOfDay >= StartMinutes || minutesOfDay < EndMinutes; // crosses midnight
    }

    public static AlarmSchedule From(Settings s) => new(
        Windowed: s.AlarmScheduleMode == Settings.ScheduleWindow,
        StartMinutes: s.AlarmWindowStartMinutes,
        EndMinutes: s.AlarmWindowEndMinutes);
}
